Fixes bug 1000160 - Set the correct version number in processed crashes.

AdrianGaudebert opened the pull request:
@twobraids Does this look good to you? 

Note that this is not ready to be merged yet. I need to figure out a transition plan, and to modify a bunch of code related to those version numbers.

twobraids commented on socorro/processor/mozilla_transform_rules.py:
```diff
@@ -816,3 +816,79 @@ def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+            if real_version:
+                processed_crash['version'] = real_version
+            elif processed_crash['release_channel'].lower() == 'beta':
+                # This is a beta version but we do not have data about it. It
```
you might want to add a note to the ```processor_meta.processor_notes``` (list of strings) to say express what happened here.

twobraids commented on socorro/processor/mozilla_transform_rules.py:
```diff
@@ -816,3 +816,79 @@ def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+                # could be because we don't have it yet (if the cron jobs are
+                # running late for example), so we mark this crash. This way,
+                # we can reprocess it later to give it the correct version.
+                processed_crash['version'] += 'b0'
```
if this crash is reprocessed for some reason other than this _unknown beta_ situation and it is still an unknown beta, then this line will tag an additional 'b0' to the end of the version.

twobraids commented on socorro/processor/mozilla_transform_rules.py:
```diff
@@ -816,3 +816,79 @@ def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+        results = transaction(
+            execute_query_fetchall,
+            sql
+        )
```
how frequently does this table change?  This class is instantiated once during processor startup so the content of this lookup table will persist unchanged until the SIGHUP or restart of the processor.

rhelmer replied:
This table is refreshed daily by crontabber ```ProductVersionsCronApp```. It may also be updated more frequently by the admin UI.

twobraids replied:
Is there another way to detect if a crash is from a beta version?  If we could detect that it is a beta version in the rule predicate then we could  do the lookup dynamically in the body of the action.  We could do a time based cache on the results to limit the number of hits on PG, but still allow it to be dynamic enough to catch changes to that table.

rhelmer replied:
We can tell if it's from *a* beta version based on the channel, we just don't have any way tell tell exactly *which* beta version this is unless we look in ```product_versions```.

AdrianGaudebert replied:
I'm not even sure this logic is needed for beta only. At the moment, we are doing this transform logic for all non-release channels in our middleware. But maybe "no release" is a good enough rule, since it will already remove a lot of crashes?

AdrianGaudebert commented:
@twobraids OK I have moved the SQL query to action time, added something to the ``processor_notes``, and added some logic change to the middleware so that nothing breaks during the transition.

AdrianGaudebert commented:
Should I add another rule, in the classification rules system, to mark all crashes with a ``b0`` suffix so they are easily re-processable? Or is it too early? Do we have the reprocessing system yet?

rhelmer commented on socorro/processor/mozilla_transform_rules.py:
```diff
@@ -816,3 +816,93 @@ def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+    def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+        try:
+            if processed_crash['release_channel'].lower() == 'release':
+                return True
```
I think you could do this for ```!= 'beta'```, why not?

rhelmer replied:
We don't have this problem with Nightly or Aurora channels, it's only Beta that send crash reports without the specific beta number. The former two are always ```a1``` and ```a2``` respectively no matter what.

twobraids replied:
this test to see if the ```_action``` should apply at all ought to be put in the ```_predicate``` method.   See the implementation https://github.com/mozilla/socorro/blob/master/socorro/processor/mozilla_transform_rules.py#L462 for an example.

AdrianGaudebert replied:
@rhelmer that's what we do in the middleware, and as far as I remember I was told to do so by Josh. Maybe he told me to do that to be future-ready, in case we started doing rapid aurora and such? 

For context in the middleware, see: https://github.com/mozilla/socorro/blob/master/socorro/external/postgresql/base.py#L400 and https://github.com/mozilla/socorro/blob/master/config/middleware.ini-dist#L472

rhelmer replied:
Hm well right now it's just a waste since we only take action on beta, I think it's not worth it.

AdrianGaudebert replied:
Good to know! I'll change that.

rhelmer commented on socorro/external/postgresql/base.py:
```diff
@@ -85,7 +85,7 @@ def query(self, sql, params=None, error_message=None, connection=None):
                 connection = self.database.connection()
                 fresh_connection = True
-            # self.context.logger.debug(connection.cursor.mogrify(sql, params))
+            # logger.debug(connection.cursor().mogrify(sql, params))
```
This is funny and I have used it before too :) Why don't we allow it to be on for ```debug```, we can run production in ```error``` and below if we don't like it there? Seems like this is what log levels are intended for.

AdrianGaudebert replied:
Historically we have always showed ``debug`` in our production logs, and that has been a pain. But we could easily change that I suppose. 

The other reason I can think about is that this is an option that you only want to turn on when you are debugging those SQL queries, which is not always the case. It's quite painful to have to go through a lot of log when debug something unrelated, so I think less log is better. 

But maybe it's just convenient to me because I know exactly where to go to turn this debugging on...

rhelmer replied:
Yeah me too (and don't let me derail this PR with this discussion!), it honestly wouldn't hurt to have this in prod either and would be helpful sometimes. If we're running with debug on in prod we kind of deserve/expect to have debug output, and this is not so much really.

twobraids commented on socorro/processor/mozilla_transform_rules.py:
```diff
@@ -816,3 +816,93 @@ def _action(self, raw_crash, raw_dumps, processed_crash, processor_meta):
+            if processed_crash['release_channel'].lower() == 'release':
+                return True
+
+            versions_data = self._get_versions_data()
```
not exactly what I had in mind, sorry for not being clearer.  

Rather than refresh the whole cached version of the table every time, change the query so that it looks up the result for this particular product/version/channel/build.  Keep a cache of the results of this query.  

When entering the ```_action```, lookup the product/version/channel/build in the cache, if it is not there, then do the query and add the results to the cache.

AdrianGaudebert commented:
OK I think this only needs a migration now. @twobraids, @rhelmer, did I forget anything?

rhelmer commented:
r+

twobraids commented:
r+

AdrianGaudebert commented on socorro/external/postgresql/raw_sql/procs/001_update_reports_clean.sql:
```diff
@@ -253,7 +253,7 @@ SET product_version_id = product_versions.product_version_id
 WHERE reports_clean_buffer.uuid = new_reports.uuid
     AND new_reports.product = product_versions.product_name
-    AND new_reports.version = product_versions.release_version
+    AND new_reports.version = product_versions.version_string
```
@selenamarie Can you review this little change please? I don't know enough about our schema to know if that will have any consequences.
